minio/minio
# Add object retention support

## Description
Implement Object locking API calls as per https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lock.html

The global worm setting represented by MINIO_WORM continues to take precedence. The previous PR #8120 added enforcement of WORM at bucket level - however the API referenced above is more granular to lock at the object level and following that here.
## Motivation and Context
PUT, HEAD, GET, DELETE and GetObjectRetention calls have been enhanced to support object locking 
Note: for PUT operation
-- Since MinIO doesnt have versioning support, if PUT tries to overwrite a object under governance, the PUT is disallowed
-- Content-MD5 sum is required for this API and is currently not enforced in this PR.

Above API's are at a point where they can be reviewed/tested.

## How to test this PR?
- Use `mc lock` to set/clear governance/compliance modes
- Use aws-cli to test remaining api calls
```
 aws s3api put-object --bucket kannappan300 --key global --object-lock-mode GOVERNANCE --object-lock-retain-until-date "2019-11-20"   --body /etc/issue --profile minio  --endpoint-url http://localhost:9000

 aws s3api get-object-retention --bucket kannappan300 --key mykey --version-id 123 --endpoint-url http://localhost:9000 --profile minio
```
Create user1 with full permissions to governance (See sample policy below). 
Create user2 will specific governance permissions removed from policy.

Testing  governance enforcement for above users would be a good test.
Also turning on site wide WORM with MINIO_WORM or changing object lock modes and testing behavior after retention period is expired.
```
{
 "version": "2012-10-17",
 "Statement": [
   {
     "Action": [
      "s3:CreateBucket",
      "s3:DeleteBucket",
      "s3:ListBucket",
      "s3:ListAllMyBuckets",
      "s3:GetBucketLocation",
       "s3:PutObject",
       "s3:GetObject",
       "s3:PutObjectRetention",
       "s3:GetObjectRetention",
       "s3:AbortMultipartUpload",
      "s3:DeleteObject",
       "s3:ListMultipartUploadParts",
       "s3:BypassGovernanceMode",
       "s3:BypassGovernanceRetention",
       "s3:GetBucketObjectLockConfiguration",
       "s3:PutBucketObjectLockConfiguration"
     ],
     "Effect": "Allow",
     "Resource": [
                  "arn:aws:s3:::*"
                 ],
     "Sid": "AllowAllActions"
   }]
}
```

## Types of changes

## Checklist:
